server/public/pluginapi/kv_memory_test.go

```diff
@@ -0,0 +1,469 @@
+package pluginapi_test
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/mattermost/mattermost/server/public/pluginapi"
+)
+
+// kvStore is used to check that KVService and MemoryStore implement the same interface.
+// Methods names are sorted alphabetically for easier comparison.
+type kvStore interface {
+	Delete(key string) error
+	DeleteAll() error
+	Get(key string, o any) error
+	ListKeys(page, count int, options ...pluginapi.ListKeysOption) ([]string, error)
+	Set(key string, value any, options ...pluginapi.KVSetOption) (bool, error)
+	SetAtomicWithRetries(key string, valueFunc func(oldValue []byte) (newValue any, err error)) error
+}
+
+var _ kvStore = (*pluginapi.MemoryStore)(nil)
+var _ kvStore = (*pluginapi.KVService)(nil)
+
+func TestMemoryStoreSet(t *testing.T) {
```

Comment:
It would be really neat if we could run the "real tests" against both implementations to ensure the tests don't diverge.

Reply:
See https://github.com/mattermost/mattermost/pull/26244#discussion_r1494982730